File: api/src/main/java/fr/huiitre/tools/modules/riot/sync/application/SyncValorantBundlesUseCase.java
package fr.huiitre.tools.modules.riot.sync.application;

import java.util.Map;
import java.util.Objects;
import java.util.Optional;
import java.util.Set;
import java.util.UUID;
import java.util.stream.Collectors;

import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import fr.huiitre.tools.modules.core.module.domain.ModuleCode;
import fr.huiitre.tools.modules.core.role.domain.RoleCode;
import fr.huiitre.tools.modules.core.security.application.usecase.SecuredUseCase;
import fr.huiitre.tools.modules.riot.valorant.application.catalog.view.ValorantBundleView;

@Service
@Transactional
public class SyncValorantBundlesUseCase implements SecuredUseCase {

    private final ValorantBundleDataProvider bundleDataProvider;
    private final ValorantBundleSyncRepository bundleSyncRepository;

    public SyncValorantBundlesUseCase(
            ValorantBundleDataProvider bundleDataProvider,
            ValorantBundleSyncRepository bundleSyncRepository) {
        this.bundleDataProvider = bundleDataProvider;
        this.bundleSyncRepository = bundleSyncRepository;
    }

    @Override
    public Optional<ModuleCode> requiredModule() {
        return Optional.of(ModuleCode.RIOT);
    }

    @Override
    public RoleCode requiredRole() {
        return RoleCode.TECH;
    }

    public ValorantSyncReport execute() {
        Map<UUID, ValorantBundleView> currentByAssetId = bundleSyncRepository.findAll().stream()
                .collect(Collectors.toMap(ValorantBundleView::assetId, it -> it));

        Set<UUID> externalAssetIds = bundleDataProvider.fetchAll().stream()
                .map(ValorantBundleSyncData::getAssetId)
                .collect(Collectors.toSet());

        int created = 0;
        int updated = 0;
        int deleted = 0;

        for (ValorantBundleSyncData ext : bundleDataProvider.fetchAll()) {
            ValorantBundleView existing = currentByAssetId.get(ext.getAssetId());

            if (existing == null) {
                bundleSyncRepository.save(ext);
                created++;
                continue;
            }

            boolean changed = !Objects.equals(existing.name(), ext.getName())
                    || !Objects.equals(existing.bannerUrl(), ext.getBannerUrl());

            if (changed) {
                bundleSyncRepository.update(existing.id(), ext);
                updated++;
            }
        }

        for (ValorantBundleView current : currentByAssetId.values()) {
            if (!externalAssetIds.contains(current.assetId())) {
                bundleSyncRepository.delete(current.id());
                deleted++;
            }
        }

        return new ValorantSyncReport(created, updated, deleted);
    }
}

File: api/Modules/Riot/Sync/Application/Usecases/SyncValorantBundlesUseCase.cs
using Tools.Api.Modules.Riot.Sync.Application.Ports;

namespace Tools.Api.Modules.Riot.Sync.Application.Usecases;

public sealed class SyncValorantBundlesUseCase(
    IValorantBundleDataProvider bundleDataProvider,
    IValorantBundleSyncRepository bundleSyncRepository
)
{
    public async Task<ValorantSyncReport> Execute()
    {
        var currentByAssetId = (await bundleSyncRepository.FindAll())
            .ToDictionary(bundle => bundle.AssetId);

        var external = await bundleDataProvider.FetchAll();
        var externalAssetIds = external.Select(bundle => bundle.AssetId).ToHashSet();

        var created = 0;
        var updated = 0;
        var deleted = 0;

        foreach (var bundle in external)
        {
            if (!currentByAssetId.TryGetValue(bundle.AssetId, out var existing))
            {
                await bundleSyncRepository.Save(bundle);
                created++;
                continue;
            }

            if (existing.Name != bundle.Name || existing.BannerUrl != bundle.BannerUrl)
            {
                await bundleSyncRepository.Update(existing.Id, bundle);
                updated++;
            }
        }

        foreach (var current in currentByAssetId.Values.Where(bundle => !externalAssetIds.Contains(bundle.AssetId)))
        {
            await bundleSyncRepository.Delete(current.Id);
            deleted++;
        }

        return new ValorantSyncReport(created, updated, deleted);
    }
}
